### DISCUSSION TITLE
How to integrate Firebase v9 with Nuxt3?

### DISCUSSION THREAD

**Discussion by akhilnarayanan1** [nuxt/nuxt#16415] [Questions] [answered]
Hi, I am trying to setup firebase v9 with nuxt3, referring this [medium article](https://lupas.medium.com/firebase-9-beta-nuxt-js-981cf3dac910)
but getting **`Component auth has not been registered yet`** error. Can anyone please help me 😢

Steps to reproduce -
1. **/assets/js/firebase.ts**
```typescript
import { initializeApp } from 'firebase/app';
import { initializeAuth, browserLocalPersistence, 
browserSessionPersistence, indexedDBLocalPersistence, getAuth } from "firebase/auth";

const firebaseConfig = {
    apiKey: "xxx",
    authDomain: "xxx",
    projectId: "xxx",
    storageBucket: "xxx",
    messagingSenderId: "xxx",
    appId: "xxx",
    measurementId: "xxx"
};

const firebaseApp = initializeApp(firebaseConfig);
const auth = initializeAuth(firebaseApp, {
    persistence: [indexedDBLocalPersistence, browserSessionPersistence, browserLocalPersistence] ,
});

export default auth;

```

2. **app.vue** -
```html
<template>
  <div>
    <NuxtPage />
  </div>
</template>

<script setup lang="ts">
import auth from '@/assets/js/firebase';

console.log(auth)
  

</script>

```

**Comment by hputzek** [comment]
I put all firebase stuff into a plugin `./plugins/firebase.client.ts`
When I looked at your code I realized you use `initializeAuth`
I use `getAuth`, but I think this is not the cause of the error:
```
initializeAuth(app: FirebaseApp, deps?: Dependencies): Auth;

This function allows more control over the {@link Auth} instance than
{@link getAuth}. `getAuth` uses platform-specific defaults to supply
the {@link Dependencies}.
```


Maybe try out if moving your code to a nuxt plugin works for you.
Here is my implementation for reference:
```typescript
import { defineNuxtPlugin, useState } from '#app'
import { initializeApp } from 'firebase/app'
import { getFirestore } from 'firebase/firestore'
import { getAuth, } from 'firebase/auth'
import firebaseConfig from '~/firebaseConfig'
// Initialize Firebase
const firebaseApp = initializeApp(firebaseConfig)
export const db = getFirestore(firebaseApp)
export const auth = getAuth(firebaseApp)

export default defineNuxtPlugin((nuxtApp) => {
    useState('firebaseApp', () => firebaseApp)
    useState('auth', () => auth)
    useState('db', () => db)
})
```

In your vue component you can access the auth instance by
```typescript
//<script setup lang="ts">
const auth = useState('auth').value
// </script>
```
(see https://v3.nuxtjs.org/docs/usage/state)

 https://v3.nuxtjs.org/docs/directory-structure/plugins

**Reply by akhilnarayanan1** [reply]
Thanks @hputzek for your reply, I tried as you suggested and I am able to get **`auth`** under app.vue.
But I get another issue - **`ERROR  [worker] Cannot read properties of undefined (reading 'onAuthStateChanged')`**,
when I try to use **`onAuthStateChanged`** in app.vue


**app.vue**-

```html
<script setup lang="ts">
  import { onAuthStateChanged, Auth } from 'firebase/auth';
  
  const auth = useState('auth').value
  
  console.log(auth)
  
  onAuthStateChanged(auth as Auth, (user) => {
    if (user) {
      console.log(user)
    } else {
      console.log("Not registered")
    }
  });
</script>
```

**Comment by hputzek** [comment, marked as answer]
This is a wild guess but have you tried to set the "onAuthState" Handler in the "mounted" hook instead of "directly" as in your code above?

```html
<script setup lang="ts">
onMounted(() => {
onAuthStateChanged(auth as Auth, (user) => {
    if (user) {
      console.log(user)
    } else {
      console.log("Not registered")
    }
  });
}
</script>
```

**Reply by akhilnarayanan1** [reply]
Yes it works, and the last thing would be to listen continuously for **`onAuthStateChanged`**. For this I am using **`useState`** in **app.vue** and calling it on **login.vue**, but I am getting useState value as undefined in login.vue


1. **app.vue**
```html
<script setup lang="ts">
    import { onAuthStateChanged, signInAnonymously, Auth } from 'firebase/auth';

    const auth = useState('auth').value

    onMounted(()=>{ 
      onAuthStateChanged(auth as Auth, (user) => {
          if (user) {
            useState('firebaseUser', () => user) // <------- creating firebaseUser state
          } else {
            anonymousSignIn()
          }
      });
    })

    const anonymousSignIn = () => {
        signInAnonymously(auth as Auth)
        .then(() => {})
        .catch((error) => {});
    }
</script>

```

2. **login.vue**
```html
<script setup lang="ts">
    console.log(useState('firebaseUser')) // <----- prints ObjectRefImpl with value as firebase user
    console.log(useState('firebaseUser').value) // <-----  prints UNDEFINED
</script>
```

**Reply by wobsoriano** [reply]
Shouldn't you be doing it like this?

```html
<script setup lang="ts">
const user = useState('firebaseUser')

// access the value via user.value
// don't access it by doing useState('firebaseUser').value

// or watch for changes
watch(user, (val) => {
  console.log('current user', val)
})
</script>
```

**Reply by akhilnarayanan1** [reply]
The `user.value` returns undefined, however the watch works perfect. Any guess why is this so?

**Reply by wobsoriano** [reply]
That's correct. It should be undefined at first until it has value. That's why `watch` works.

```html
<script setup lang="ts">
const user = useState('firebaseUser')
</script>

<template>
  <div v-if="user">Authenticated: {{ JSON.stringify(user) }}</div>
  <div v-else>Not authenticated</div>
</template>
```

**Reply by hputzek** [reply]
_Its in general not a good idea to put logic meant to be reused directly into a vue component._

If you want to use the "getUser" functionality in several components you should either use vuex or put the logic to get the user into a composable in order to reuse it. ([see composables docs](https://v3.nuxtjs.org/docs/directory-structure/composables))


Here is a working example I just put together:

`./plugins/firebase.client.ts`
```typescript
import { defineNuxtPlugin, useState } from '#app'
import { initializeApp } from 'firebase/app'
import { getFirestore } from 'firebase/firestore'
import { getAuth, } from 'firebase/auth'
import firebaseConfig from '~/firebaseConfig'
// Initialize Firebase
const firebaseApp = initializeApp(firebaseConfig)
export const db = getFirestore(firebaseApp)
export const auth = getAuth(firebaseApp)

export default defineNuxtPlugin((nuxtApp) => {
    useState('firebaseApp', () => firebaseApp)
    useState('auth', () => auth)
    useState('db', () => db)
})
```

`./composables/useGetCurrentUser.ts`
```typescript
import { onAuthStateChanged } from 'firebase/auth'

const getUserData = () => {
    const user = ref(null)
    const auth = useState('auth').value
    onMounted(() => {
        onAuthStateChanged(auth, (currentUser) => {
            if (currentUser) {
                user.value = currentUser
            }
        })
    })
    return user
}

export const useGetUserData = () => {
    return getUserData
}
```

`test.vue` -> can be any vue component
```html
<template>
    <pre>{{ JSON.stringify(user, null, 2) }}</pre>
</template>
<script setup>
import { useGetUserData } from '~/composables/useGetCurrentUser'
const getUserData = useGetUserData()
const user = getUserData() // this is a reactive property since the composable returns a `ref`
</script>
```

**Reply by akhilnarayanan1** [reply]
Thanks @wobsoriano , everything is working fine for me now 😄

**Reply by hputzek** [reply]
Ok, nice to hear! - Maybe have a look at composables anyways, could be helpful

**Reply by jrsamuel** [reply]
This was very helpful.  However, I would like to have access to the user before Vue initializes so that I can effect the routing (e.g. don't let non-authenticated users go to pages that require auth).  Is there a way to start the onAuthStateChanged listener so that it impacts routing?

**Reply by hputzek** [reply]
> before Vue initializes

how are you going to do any routing before vue initializes?

> Is there a way to start the onAuthStateChanged listener so that it impacts routing?

Yes, the `./composables/useGetCurrentUser.ts` example from my last comment works without vue, too:
You can put this in any js file:
```javascript
import { useGetUserData } from '~/composables/useGetCurrentUser'
const getUserData = useGetUserData()
const user = getUserData() // this is a reactive property since the composable returns a `ref`
```
> don't let non-authenticated users go to pages that require auth

You can e.g hook into vue router to do this - [Here is a tutorial.](https://dev.to/gautemeekolsen/vue-guard-routes-with-firebase-authentication-f4l)
Be aware: if you do not use ssr but have all the routing/auth logic clientside it's not 100% secure - because in theory somebody could just alter your JS to allow the routes again.

**Reply by Utitofon-Udoekong** [reply]
Is there any workaround if you are using the firebase admin sdk since everything is now on the server and not the client side?

**Reply by dev19-softrix** [reply]
import { initializeApp } from "firebase/app";
import { getFirestore } from 'firebase/firestore';


export const useFirebase = async () => {
    const firebaseConfig = {
        //
    };

    const firebaseApp = initializeApp(firebaseConfig);
    const firestore = getFirestore(firebaseApp);
    return {
        firebaseApp,
        firestore,
    }
}

**Comment by wobsoriano** [comment]
You can also use `provide` from `nuxtApp` when creating a plugin

```ts
// plugins/firebase.client.ts
import { defineNuxtPlugin } from '#app';

import { FirebaseApp, initializeApp } from 'firebase/app';
import { Auth, getAuth } from 'firebase/auth';

declare module '#app' {
  interface NuxtApp {
    $firebaseApp: FirebaseApp;
    $firebaseAuth: Auth;
  }
}

export default defineNuxtPlugin((nuxtApp) => {
  const config = useRuntimeConfig();
  const firebaseConfig = {
    apiKey: config.firebaseApiKey,
    authDomain: config.firebaseAuthDomain,
  };

  const app = initializeApp(firebaseConfig);

  nuxtApp.provide('firebaseApp', app);
  nuxtApp.provide('firebaseAuth', getAuth(app));
});
```

```ts
// nuxt.config.ts
import { defineNuxtConfig } from 'nuxt3';

export default defineNuxtConfig({
  publicRuntimeConfig: {
    firebaseApiKey: process.env.FIREBASE_API_KEY,
    firebaseAuthDomain: process.env.FIREBASE_AUTH_DOMAIN,
  },
});
```

```html
<script setup lang="ts">
import { signInWithEmailAndPassword } from 'firebase/auth';

const { $firebaseAuth } = useNuxtApp();

onMounted(async () => {
  await signInWithEmailAndPassword(
    $firebaseAuth,
    'email@gmail.com',
    'password'
  );
});
</script>
```

**Comment by jrsamuel** [comment]
I am continuing to play around with the best "Nuxt 3" way of handing auth with Firebase.  I came up with this approach that uses a simple Pinia store for the user data so you don't need a separate composable.

 In my `firebase.client.ts` plugin I access the Pinia store, wrap the 'onAuthStateChanged' listener in a Promise and call and await that function which updates the UserStore with the current user.  Awaiting the Promise ensures that  the listener callback finishes so that I have the current user state even on page refresh.   You just need to access the Pinia store from your components.

Lastly, I use `addRouteMiddleware` to set up a route guard that has access to the current user.  This can be modified based on the needs of the project.

```
... 

// Initialize Firebase
initializeApp(firebaseConfig)

export default defineNuxtPlugin(async (nuxtApp) => {
    const UserState = useStore()
    const startAuthListener = () => {
        return new Promise((resolve, reject) => {
            onAuthStateChanged(getAuth(), (user: User | null) => {
                UserState.user = user
                resolve(null)
            })
        })
    }

    await startAuthListener()
    addRouteMiddleware('auth', (to, from) => {
        if (!UserState.user && to.path === '/private') {
            return navigateTo('/')
        }
    }, { global: true })
})
```

This is still a work in progress so if you have any suggestions/improvements please let me know.  For instance, I'm not sure what is the best way to prevent access to restricted pages if sometimes types the restricted page url directly?  In that case, this route guard doesn't trigger.

**Comment by wobsoriano** [comment]
Made a starter for authenticated server-side rendering with Nuxt 3 and Firebase 9.

https://github.com/wobsoriano/nuxt-firebase-auth-starter

**Reply by areindl** [reply]
Thanks to @antfu 

This is needed for Nuxt3 RC3 so that Auth is loading correctly.

```js
// nuxt.config.ts
export default defineNuxtConfig({
  experimental: {
    viteNode: true
  }
```
})

Source: [https://github.com/nuxt/framework/issues/4584#issuecomment-1123340177](https://github.com/nuxt/framework/issues/4584#issuecomment-1123340177)

**Comment by dosstx** [comment]
Does anyone know how to detach a firestore listener in nuxt3? I am attaching it via the `onMounted()` hook like so:

```
const unsubscribe = ref(null) //<--- hold it here?

onMounted(async () => {
  const { firestore } = useFirebase()
  const userPostsRef = collection(firestore, 'posts')
  const q = query(userPostsRef, where('uid', '==', userStore.userProfile.uid))
  unsubscribe.value = onSnapshot(q, (snapshot) => { // <---start listener here
    snapshot.docChanges().forEach((change) => {
      if (change.type === 'added') {
        console.log('added', change.doc.data())
        userPosts.push(change.doc.data())
      }
....snip...
})
}
```

and then later in the component I have the `onUnMounted()` hook like so:

```
onUnmounted(() => {
  unsubscribe.value //<----detach it here??
})
```
**But this doesn't "work" because I still get console logs from my listener throughout the app.**

**Reply by wobsoriano** [reply]
Should be unsubscribe.value() ?

**Comment by dosstx** [comment]
Does anyone have an example of using Nuxt3's `useFetch` composable and fetching firestore data from `server/api` route ? I have a set up that works well , but when the user reloads the page, I get a 500 error:
```js
[nuxt] [request error] Cannot read properties of undefined (reading 'content')
  at _sfc_ssrRender (/C:/Users/XXX/Desktop/work/XXX/.nuxt/dist/server/server.mjs:36448:32)
  ```

Here's my question in detail: https://stackoverflow.com/questions/73291222/nuxt3-why-cannot-read-properties-of-undefined-not-sure-if-hydration-error

**Comment by dosstx** [comment]
Does anyone have .env file working with Nuxt3 + Firebase serverless hosting? I got my app to work fine locally, but when deploying to Firebase Hosting, I am having issues with the app reading my .env file. I am deploying via Firebase cloud functions and it's not reading my environment variables.

**Reply by AmitGurbani** [reply]
Did you try setting env variables for firebase functions?
`firebase functions:config:set config.key="SECRET_KEY" config.pass="SECRET_PASS"`

https://dev.to/rajeshkumaravel/google-firebase-functions-setting-and-accessing-environment-variable-1gn2

**Comment by dosstx** [comment]
Did firebase hosting stop working for you guys? For some reason I can no longer deploy to Firebase Hosting with Cloud functions. I get the below error:

```js
i  functions: updating Node.js 16 function nuxt:server(us-central1)...
Build failed: npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR!
npm ERR! While resolving: undefined@undefined
npm ERR! Found: firebase-functions@3.24.1
npm ERR! node_modules/firebase-functions
npm ERR!   firebase-functions@"3.24.1" from the root project
npm ERR!
npm ERR! Could not resolve dependency:
npm ERR! peer firebase-functions@"^4.0.0" from firebase-functions-test@3.0.0
npm ERR! node_modules/firebase-functions-test
npm ERR!   firebase-functions-test@"latest" from the root project
npm ERR!
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps


Functions deploy had errors with the following functions:
        nuxt:server(us-central1)

Error: There was an error deploying functions
```
